### PR TITLE
#169258632 Login in to Barefoot application via Facebook and Google

### PR DESCRIPTION
**Pull request by victkarangwa**
#### What does this PR do?
This PR allows the user to login with social accounts
#### Description of Task to be completed?
- the user is able to login with Facebook account
- the user is able to login with Google account
#### How should this be manually tested?
1.  `clone` the repo
2.  Run `npm install` all required dependencies
3. Run `npm run dev-start` to start the server
4. Open postman application and navigate to:
  **For facebook**: `post localhost:3000/api/auth/facebook/`
  **For google**: `post localhost:3000/api/auth/google/`
5. Inside body write `"access_token": "[your_access_token]"` then hit **send**

#### Any background context you want to provide?
Previously, users were required to first create a new account on the platform for them to be able to login
#### What are the relevant pivotal tracker stories?
[#169258632](https://www.pivotaltracker.com/story/show/169258632)
##### Signin with facebook
Screen Shot 2019-11-13 at 16 23 30

##### In the database
Screen Shot 2019-11-13 at 16 41 08

##### Signin with google
Screen Shot 2019-11-13 at 16 25 12

##### In the database
Screen Shot 2019-11-13 at 16 41 25

##### swagger documentation
Screen Shot 2019-11-13 at 19 43 31

#### Questions:
N/A

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -8,49 +8,63 @@
     "build": "babel src -d dist ",
     "eslint": "eslint --ignore-path .gitignore .",
     "dev": "nodemon --exec babel-node ./src/index.js",
-    "test": "NODE_ENV=testing nyc mocha ./src/tests/*.js  --timeout 200000 --exit",
-    "coverage": "nyc npm test  && npm run coveralls-coverage"
+    "test": "NODE_ENV=testing nyc --reporter=text --reporter=html --require @babel/polyfill --require @babel/register mocha ./src/tests/controllers/*.js  --timeout 200000 --exit",
+    "coverage": "nyc npm test  && npm run coveralls-coverage",
+    "create-db": "npx sequelize-cli model:generate --name users --attributes firstName:string,lastName:string,email:string,password:string,isVerified:boolean,signupType:string && sequelize db:migrate",
```
Comment (william000000) [inline package.json]: It's good you changed this!

File: src/helpers/authenticateToken.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
+
+import jwt from 'jsonwebtoken';
+import dotenv from 'dotenv';
+
+dotenv.config();
+/**
+ * @export
+ * @class Helper
+ */
+export default class AuthToken {
```
Comment (william000000) [inline src/helpers/authenticateToken.js]: Rename this file as your class name is, use Pascal Case for the class

**Review by william000000 — CHANGES_REQUESTED** [review]

File: .env.example [hunk 3]
```diff
@@ -1,2 +1,8 @@
-DATABASE_URL_DEV=''
-DATABASE_URL_TEST=''
+DATABASE_DEV_URL=''
+DATABASE_TEST_URL=''
+passwordEnv=''
+JWT_KEY=''
+FACEBOOK_APP_ID=''
+FACEBOOK_APP_SECRET=''
+GOOGLE_APP_ID=''
+GOOGLE_APP_SECRET=''
```
Comment (julietezekwe) [inline .env.example]: End of line

File: src/controllers/userController.js [hunk 4]
```diff
@@ -0,0 +1,122 @@
+import HttpStatus from 'http-status-codes';
+import sendResult from '../utils/sendResult';
+import tokenGenerator from '../helpers/AuthToken';
+import { users } from '../database/models';
+/**
+ * @exports
+ * @class userController
+ */
+class userController {
```
Comment (julietezekwe) [inline src/controllers/userController.js]: Please follow the right naming convention

Comment (julietezekwe) [reply src/controllers/userController.js]: Use PascalCase for Class naming and the file name as well

File: src/database/config/config.js [hunk 5]
```diff
@@ -1,22 +1,18 @@
...
+    "use_env_variable": "DATABASE_URL",
+    "dialect": "postgresql"
   }
-  
-  module.exports = config;
-  
\ No newline at end of file
+}
+
+module.exports = config;
```
Comment (julietezekwe) [inline src/database/config/config.js]: End of line

File: src/database/migrations/20191113080359-create-users.js [hunk 6]
```diff
@@ -30,4 +39,4 @@ module.exports = {
   down: (queryInterface, Sequelize) => {
     return queryInterface.dropTable('users');
   }
-};
+};
```
Comment (julietezekwe) [inline src/database/migrations/20191113080359-create-users.js]: End of line

File: src/database/models/users.js [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+    email: DataTypes.STRING,
+    password: DataTypes.STRING,
+    isVerified: DataTypes.BOOLEAN,
+    signupType: DataTypes.STRING
+  }, {});
+  users.associate = function(models) {
+    // associations can be defined here
+  };
+  return users;
+};
```
Comment (julietezekwe) [inline src/database/models/users.js]: End of line

File: src/routes/api/index.js [hunk 8]
```diff
@@ -1,18 +1,8 @@
-const router = require("express").Router();
+import { Router } from 'express';
+import authRoute from './socialRoute';
```
Comment (julietezekwe) [inline src/routes/api/index.js]: The import name should be the same as the file name. Here should be import socialRoute from '/socialRoute'

File: src/routes/api/index.js [hunk 9]
```diff
@@ -1,18 +1,8 @@
...
 
-router.use("/", require("./users"));
+const router = Router();
+router.use('/auth', authRoute);
```
Comment (julietezekwe) [inline src/routes/api/index.js:7]: Then change the authRoute to SocialRoute

**Review by julietezekwe — CHANGES_REQUESTED** [review]
Great Job @victorkarangwa4. Please attend to the comments

**Review by hezronkimutai — APPROVED** [review]
Great work vicky, We are good to go.

File: src/database/migrations/20191113080359-create-users.js (hunk 6 above)
Comment (julietezekwe) [reply src/database/migrations/20191113080359-create-users.js]: @victorkarangwa4 Please attend to this

Comment (julietezekwe) [reply src/database/migrations/20191113080359-create-users.js]: Or resolve the conversation if you have implemented it

**Review by julietezekwe — COMMENTED** [review]
Please fix the end of line I pointed out and get those who requested changes on your PR to re-review it so we can merge

File: src/utils/sendResult.js [hunk 10]
```diff
@@ -0,0 +1,9 @@
+import HttpStatus from 'http-status-codes';
+
+const sendResult = async (res, status, msg, data) => {
+  if (data && data.length === 0) {
+    return res.status(status).json({ status: HttpStatus.NOT_FOUND, msg: 'Oops no results found' });
+  }
+  return res.status(status).json({ status: HttpStatus.OK, msg, data });
```
Comment (NiyoEric) [inline src/utils/sendResult.js]: @victorkarangwa4 remember to use customize helper class to return all successful and error responses.

File: src/index.js [hunk 11]
```diff
@@ -9,6 +9,7 @@ app.use(bodyParser.json());
 app.use(bodyParser.urlencoded({ extended: false }));
 app.use(bodyParser.text());
 app.use(bodyParser.json({ type: 'application/json' }));
+app.use('/', router);
 app.use(express.json());
 
 app.use('/', router);
```
Comment (NiyoEric) [inline src/index.js]: @victorkarangwa4 remember to remove duplicate `app.use('/', router);`

**Review by NiyoEric — COMMENTED** [review]
Good job @victorkarangwa4 and remember to implement feedback.

File: src/utils/sendResult.js (hunk 10 above)
Comment (victorkarangwa4) [reply src/utils/sendResult.js]: Thanks @NiyoEric for the feedback
This function acts like a customizer

**Review by william000000 — APPROVED** [review]

**Review by NiyoEric — APPROVED** [review]

**Review by NiyoEric — APPROVED** [review]

**Review by NiyoEric — APPROVED** [review]

**Review by ivymwende — APPROVED** [review]
@victorkarangwa4 Great job on implementing your feature. Your code is modular in a good way keep it up!

**Review by julietezekwe — APPROVED** [review]

**Pull request merged**